v2/health.go

```diff
@@ -0,0 +1,146 @@
+	options *HealthCheckerOptions
+}
+
+// HealthCheckerOptions configures the HealthChecker.
+// HealthCheckInterval defaults to 1 minute if not set or set to <= 0.
+// HealthCheckTimeout defaults to HealthChecker.interval if not set or set to 0 or set to be larger than interval.
+type HealthCheckerOptions struct {
+	// HealthCheckInterval is the time between health checks.
+	HealthCheckInterval time.Duration
+	// HealthCheckTimeout is the context timeout for each health check
+	HealthCheckTimeout time.Duration
+}
+
+// NewHealthChecker creates a new HealthChecker with the provided clients.
+// clients is a map of namespaceName name to azservicebus.Client.
+func NewHealthChecker(clients map[string]*azservicebus.Client, options *HealthCheckerOptions) *HealthChecker {
+	if options == nil {
+		options = &HealthCheckerOptions{}
+	}
+	if options.HealthCheckInterval <= 0 {
+		options.HealthCheckInterval = defaultHealthCheckInterval
+	}
+	if options.HealthCheckTimeout <= 0 || options.HealthCheckTimeout > options.HealthCheckInterval {
+		options.HealthCheckTimeout = options.HealthCheckInterval
+	}
+
+	return &HealthChecker{
+		clients: clients,
+		options: options,
+	}
+}
+
+// StartPeriodicHealthCheck starts the periodic health check for the provided HealthCheckable.
+// The health check will run on each client in the HealthChecker.
+// Stops when the context is canceled.
+func (h *HealthChecker) StartPeriodicHealthCheck(ctx context.Context, hc HealthCheckable) {
+	for namespaceName, client := range h.clients {
+		go func(namespaceName string, client *azservicebus.Client) {
+			h.periodicHealthCheck(ctx, hc, namespaceName, client)
+		}(namespaceName, client)
+	}
+}
+
+func (h *HealthChecker) periodicHealthCheck(ctx context.Context, hc HealthCheckable, namespaceName string, client *azservicebus.Client) {
+	nextCheck := time.Now()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(time.Until(nextCheck)):
+			sbCtx, cancelFunc := context.WithTimeout(ctx, h.options.HealthCheckTimeout)
+			err := hc.HealthCheck(sbCtx, client, cancelFunc)
+			if err != nil {
+				getLogger(ctx).Error(err.Error())
+			}
+			hc.IncHealthCheckMetric(namespaceName, err)
+			nextCheck = nextCheck.Add(h.options.HealthCheckInterval)
+		}
+	}
+}
+
+// HealthCheckable is an interface for performing health checks on azservicebus.Sender and azservicebus.Receiver.
+type HealthCheckable interface {
+	HealthCheck(ctx context.Context, client *azservicebus.Client, cancelFunc context.CancelFunc) error
+	IncHealthCheckMetric(namespaceName string, healthCheckErr error)
+}
+
+// SenderHealthChecker performs health checks on azservicebus.Sender.
+type SenderHealthChecker struct {
+	EntityName string
+}
+
+// HealthCheck performs a health check on azservicebus.Sender by creating a new message batch.
+func (s *SenderHealthChecker) HealthCheck(ctx context.Context, client *azservicebus.Client, cancelFunc context.CancelFunc) error {
+	defer cancelFunc()
```

Comment:
It is unusual to pass down a context and a cancelfunc. I'd consider that a smell.
I think in this case, you can call defer cancel from the caller and it's going to be same.